### PR TITLE
kubeadm: Make it possible to configure volume mounts via the config file

### PR DESCRIPTION
**Pull request by andrewrynhard**
**What this PR does / why we need it**:
Kubeadm mounts host CA certs into api server and controller manager. It uses `/etc/pki` and does not allow for the path to be configurable. This PR adds a default to `/etc/pki` but also allows a user to configure the path in the config file. In the case of using Container Linux, the CAs are located at `/usr/share/ca-certificates`, so without this PR the hardcoded `/etc/pki` path is used and will break, for example, the `--cloud-provider` flag because of missing CAs.

Fixes https://github.com/kubernetes/kubeadm/issues/484
Fixes https://github.com/kubernetes/kubeadm/issues/476
Fixes https://github.com/kubernetes/kubeadm/issues/441

/cc @luxas

### CODE REVIEW

**Comment by andrewrynhard** [comment]
Manifests are generated correctly
```
core # cat /etc/kubernetes/manifests/kube-apiserver.yaml | grep /usr/share/ca-certificates
    - mountPath: /usr/share/ca-certificates
      path: /usr/share/ca-certificates
core # cat /etc/kubernetes/manifests/kube-controller-manager.yaml | grep /usr/share/ca-certificates
    - mountPath: /usr/share/ca-certificates
      path: /usr/share/ca-certificates
```

**Review by luxas — CHANGES_REQUESTED** [review]
I talked to @andrewrynhard about this on Slack and we decided that a generic variant of this is much better. Something like this:

```go
type HostPathMount struct {
	Name string
	HostPath string
	MountPath string
}

type MasterConfiguration struct {
        ...
	APIServerExtraVolumes []HostPathMount
	ControllerManagerExtraVolumes []HostPathMount
	SchedulerExtraVolumes []HostPathMount
}
```

**Comment by timothysc** [comment]
@mattmoyer

File: cmd/kubeadm/app/phases/controlplane/volumes.go [hunk 1]
```diff
@@ -37,11 +37,6 @@ const (
 	kubeConfigVolumeName = "kubeconfig"
 )
 
-// caCertsPkiVolumePath specifies the path that can be conditionally mounted into the apiserver and controller-manager containers
-// as /etc/ssl/certs might be a symlink to it. It's a variable since it may be changed in unit testing. This var MUST NOT be changed
-// in normal codepaths during runtime.
-var caCertsPkiVolumePath = "/etc/pki"
```
Comment (mattmoyer) [inline cmd/kubeadm/app/phases/controlplane/volumes.go]: It looks like the unit tests that used this variable need a fix ([test log](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/49840/pull-kubernetes-unit/44682/?log#log)): 
```
[...]
W0804 00:30:36.473] # k8s.io/kubernetes/cmd/kubeadm/app/phases/controlplane
W0804 00:30:36.474] cmd/kubeadm/app/phases/controlplane/volumes_test.go:276: not enough arguments in call to getEtcdCertVolumes
W0804 00:30:36.474] 	have (kubeadm.Etcd)
W0804 00:30:36.475] 	want (kubeadm.Etcd, string)
W0804 00:30:36.475] cmd/kubeadm/app/phases/controlplane/volumes_test.go:514: undefined: caCertsPkiVolumePath
W0804 00:30:36.475] cmd/kubeadm/app/phases/controlplane/volumes_test.go:515: undefined: caCertsPkiVolumePath
[...]
```

**Review by mattmoyer — CHANGES_REQUESTED** [review]
I had a couple of style questions. Other than picking the right name and fixing some tests, I don't see a reason not to make this configurable.

Distros are [pretty inconsistent](https://www.happyassassin.net/2015/01/12/a-note-about-ssltls-trusted-certificate-stores-and-platforms/) about where they keep "system" trust stores (`/etc/pki/` vs. `/usr/share/ca-certificates` vs. `/etc/ssl/`).

**Comment by mattmoyer** [comment]
I'm also +1 on the idea of a generic `HostPathMount` configuration.

**Comment by luxas** [comment]
ping @andrewrynhard You're gonna take a look at this soon, right?

File: cmd/kubeadm/app/phases/controlplane/volumes.go [hunk 2]
```diff
@@ -86,37 +86,52 @@ func getHostPathVolumesForTheControlPlane(cfg *kubeadmapi.MasterConfiguration) c
...
+		mounts.NewHostPathMount(kubeadmconstants.KubeScheduler, cfg.SchedulerExtraVolumes.Name, cfg.SchedulerExtraVolumes.HostPath, cfg.SchedulerExtraVolumes.MountPath, true, &hostPathDirectoryOrCreate)
+	}
 	return mounts
 }
 
 // controlPlaneHostPathMounts is a helper struct for handling all the control plane's hostPath mounts in an easy way
 type controlPlaneHostPathMounts struct {
-	volumes      map[string][]v1.Volume
-	volumeMounts map[string][]v1.VolumeMount
+	volumes      map[string]map[string]v1.Volume
```
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/volumes.go]: is there any simple way we can make this more readable? At least document what strings should be where?

File: cmd/kubeadm/app/util/staticpod/utils.go [hunk 3]
```diff
@@ -102,6 +102,28 @@ func NewVolumeMount(name, path string, readOnly bool) v1.VolumeMount {
 	}
 }
 
+// VolumeMapToSlice returns a slice of volumes from a map's values
+func VolumeMapToSlice(volumes map[string]v1.Volume) []v1.Volume {
```
Comment (luxas) [inline cmd/kubeadm/app/util/staticpod/utils.go]: unit tests here please (for both instances)

**Review by luxas — COMMENTED** [review]
Generally LGTM, please split up in two commits though; one with real code and one with auto-generated code

Thanks!

File: cmd/kubeadm/app/phases/controlplane/volumes.go (hunk 2 above)
Comment (andrewrynhard) [reply cmd/kubeadm/app/phases/controlplane/volumes.go]: Agreed. Really ugly. I can fix this up

**Comment by andrewrynhard** [comment]
@luxas We should document the default mounts, and their names so that users know exactly what to name a volume should they want to overwrite the default.

File: cmd/kubeadm/app/util/staticpod/utils.go (hunk 3 above)
Comment (andrewrynhard) [reply cmd/kubeadm/app/util/staticpod/utils.go]: will do

**Comment by luxas** [comment]
> @luxas We should document the default mounts, and their names so that users know exactly what to name a volume should they want to overwrite the default. Where can I do that?

http://kubernetes.io/docs/admin/kubeadm?

File: cmd/kubeadm/app/util/staticpod/utils_test.go [hunk 4]
```diff
@@ -198,6 +201,32 @@ func TestNewVolumeMount(t *testing.T) {
 		}
 	}
 }
+func TestVolumeMapToSlice(t *testing.T) {
+	testVolumes := map[string]v1.Volume{
+		"test": v1.Volume{},
```
Comment (luxas) [inline cmd/kubeadm/app/util/staticpod/utils_test.go]: a couple of more tests please?

File: cmd/kubeadm/app/util/staticpod/utils_test.go [hunk 5]
```diff
@@ -198,6 +201,32 @@ func TestNewVolumeMount(t *testing.T) {
...
+		t.Errorf("Expected slice length of 1, got %d", len(volumeSlice))
+	}
+	if volumeSlice[0].Name != "foo" {
+		t.Errorf("Expected volume name \"foo\", got %d", volumeSlice[0].Name)
+	}
+}
+func TestVolumeMountMapToSlice(t *testing.T) {
+	testVolumeMounts := map[string]v1.VolumeMount{
+		"foo": v1.VolumeMount{
+			Name: "foo",
```
Comment (luxas) [inline cmd/kubeadm/app/util/staticpod/utils_test.go]: more test cases please?

File: cmd/kubeadm/app/util/staticpod/utils_test.go [hunk 6]
```diff
@@ -198,6 +201,32 @@ func TestNewVolumeMount(t *testing.T) {
...
+		"test": v1.Volume{},
+	}
+	volumeSlice := VolumeMapToSlice(testVolumes)
+	if len(volumeSlice) != 1 {
+		t.Errorf("Expected slice length of 1, got %d", len(volumeSlice))
+	}
+	if volumeSlice[0].Name != "foo" {
+		t.Errorf("Expected volume name \"foo\", got %d", volumeSlice[0].Name)
+	}
+}
```
Comment (luxas) [inline cmd/kubeadm/app/util/staticpod/utils_test.go]: nit: one empty line between these

File: cmd/kubeadm/app/phases/controlplane/volumes.go (hunk 2 above)
Comment (luxas) [reply cmd/kubeadm/app/phases/controlplane/volumes.go]: not sure I see the comment improvement ;)

File: cmd/kubeadm/app/phases/controlplane/volumes.go [hunk 7]
```diff
@@ -86,37 +86,53 @@ func getHostPathVolumesForTheControlPlane(cfg *kubeadmapi.MasterConfiguration) c
...
-	c.volumeMounts[component] = append(c.volumeMounts[component], volMounts...)
+	for _, v := range vols {
+		c.volumes[component][v.Name] = v
+	}
+	for _, v := range volMounts {
+		c.volumeMounts[component][v.Name] = v
+	}
+}
+
+func (c *controlPlaneHostPathMounts) AddExtraHostPathMounts(component string, extraVols []kubeadmapi.HostPathMount, readOnly bool, hostPathType *v1.HostPathType) {
```
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/volumes.go]: hmm, please add comments to all exported funcs in this file.

Comment (luxas) [reply cmd/kubeadm/app/phases/controlplane/volumes.go]: also, it's intended that extravolumes take precedence over defaults,  right? Then comment that and add unit tests thanks

**Review by luxas — COMMENTED** [review]
Some more comments. It seems like the last rebase was unsuccessful, please rebase again and just include your two commits. Thanks!

**Review by luxas — APPROVED** [review]
/lgtm

Thanks for this contribution, this is very valuable to have!

**Review by luxas — APPROVED** [review]
/lgtm

**Comment by ArchiFleKs** [comment]
Hi, does this PR allows to use volume type ? I'm trying with kubeadm config file to mount the cloud_config file for an OpenStack provider but all the extra volumes are of type `DirectoryOrCreate`

**Comment by luxas** [comment]
@ArchiFleKs Not right now, please file an issue with more details if that's something you'd like to see

**Comment by usernkey** [comment]
Is the feature of volume mounts via the config file been released on the current kubeadm as I would need if I would need to mount the cloud_config to add the AWS support :)

root@ip-10-1-18-136:~# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.2", GitCommit:"81753b10df112992bf51bbc2c2f85208aad78335", GitTreeState:"clean", BuildDate:"2018-04-27T09:10:24Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

**Comment by andrewrynhard** [comment]
@usernkey it is, see https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go#L93-L99

**Comment by usernkey** [comment]
cool thanks

**Pull request merged**